teerex/src/lib.rs

```diff
@@ -162,70 +169,143 @@ pub mod pallet {
 		// the integritee-service wants to register his enclave
 		#[pallet::call_index(0)]
 		#[pallet::weight((<T as Config>::WeightInfo::register_ias_enclave(), DispatchClass::Normal, Pays::Yes))]
-		pub fn register_ias_enclave(
+		pub fn register_sgx_enclave(
 			origin: OriginFor<T>,
-			ra_report: Vec<u8>,
-			worker_url: Vec<u8>,
+			proof: Vec<u8>,
+			worker_url: Option<Vec<u8>>,
+			attestation_method: SgxAttestationMethod,
 		) -> DispatchResultWithPostInfo {
-			log::info!("teerex: called into runtime call register_ias_enclave()");
+			log::info!("teerex: called into runtime call register_sgx_enclave()");
 			let sender = ensure_signed(origin)?;
-			ensure!(ra_report.len() <= MAX_RA_REPORT_LEN, <Error<T>>::RaReportTooLong);
-			ensure!(worker_url.len() <= MAX_URL_LEN, <Error<T>>::EnclaveUrlTooLong);
+			ensure!(proof.len() <= SGX_RA_PROOF_MAX_LEN, <Error<T>>::RaProofTooLong);
+			ensure!(
+				worker_url.clone().unwrap_or_default().len() <= MAX_URL_LEN,
+				<Error<T>>::EnclaveUrlTooLong
+			);
 			log::info!("teerex: parameter length ok");
 
-			#[cfg(not(feature = "skip-ias-check"))]
-			let enclave = Self::verify_report(&sender, ra_report)?.with_url(worker_url.clone());
+			let enclave = match attestation_method {
+				SgxAttestationMethod::Ias => {
+					let report = sgx_verify::verify_ias_report(&proof)
+						.map_err(|_| <Error<T>>::RemoteAttestationVerificationFailed)?;
+					log::info!("teerex: IAS report successfully verified");
+					let enclave = SgxEnclave::new(
+						report.report_data,
+						report.mr_enclave,
+						report.mr_signer,
+						report.timestamp,
+						report.build_mode,
+						report.status,
+					)
+					.with_attestation_method(SgxAttestationMethod::Ias);
+
+					ensure!(
+						Ok(sender.clone()) ==
+							T::AccountId::decode(&mut report.report_data.lower32().as_ref()),
+						<Error<T>>::SenderIsNotAttestedEnclave
+					);
+
+					// TODO: activate state checks as soon as we've fixed our setup #83
+					// ensure!((report.status == SgxStatus::Ok) | (report.status == SgxStatus::ConfigurationNeeded),
+					//     "RA status is insufficient");
+					// log::info!("teerex: status is acceptable");
+
+					Self::ensure_timestamp_within_24_hours(report.timestamp)?;
```

Comment:
Intuitively, it makes more sense if this is directly after verifying the report.